Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift

```diff
@@ -561,34 +566,140 @@ final class AppDelegateViewModelTests: TestCase {
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.notDetermined)
 
-      self.registerUserNotificationSettings.assertValueCount(3)
+      self.getNotificationAuthorizationStatus.assertValueCount(2)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(1)
+      self.unregisterForRemoteNotifications.assertValueCount(0)
+
+      self.vm.inputs.notificationAuthorizationCompleted()
+
+      self.getNotificationAuthorizationStatus.assertValueCount(3)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(1)
+      self.unregisterForRemoteNotifications.assertValueCount(0)
+
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.authorized)
+
+      self.getNotificationAuthorizationStatus.assertValueCount(3)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(2)
       self.unregisterForRemoteNotifications.assertValueCount(0)
     }
 
     self.vm.inputs.userSessionEnded()
 
-    self.registerUserNotificationSettings.assertValueCount(3)
+    self.registerForRemoteNotifications.assertValueCount(2)
     self.unregisterForRemoteNotifications.assertValueCount(1)
   }
 
+  fileprivate func count(_ string: String) -> ([String]) -> Int {
+    return { array in
+      return array.reduce(0, { acc, val in val == string ? acc + 1 : acc })
+    }
+  }
+
+  func testTrackingPushAuthorizationOptIn() {
+    let client = MockTrackingClient()
+
+    withEnvironment(currentUser: .template, koala: Koala(client: client)) {
+      self.vm.inputs.userSessionStarted()
+
+      XCTAssertEqual(0, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.applicationDidEnterBackground()
+      self.vm.inputs.applicationWillEnterForeground()
+
+      XCTAssertEqual(0, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.notDetermined)
+
+      XCTAssertEqual(1, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.notificationAuthorizationCompleted()
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.authorized)
+
+      XCTAssertEqual(1, client.events |> count("Triggered Push Permissions Dialog"))
```

Comment:
@stephencelis, @theginbin 
First of all, thank you for looking in to it - it helps me a lot!
Second, FYI we head similar discussion with @justinswart and I was trying to convince him why I did it this particular way :). And we decided that I will keep it and submit it for the review and then we would have discussion together :) 

If I understand  the issue - you would prefer to see something like this: 
`XCTAssertEqual(["Triggered Push Permissions Dialog"], client.events))` 
instead of this:
`XCTAssertEqual(1, client.events |> count("Triggered Push Permissions Dialog"))`

My argument against `XCTAssertEqual(["Triggered Push Permissions Dialog"], client.events))` is this:

We pass tracking client to our environment before we start "manipulating" our AppDelegateViewModel. During sequence of manipulations on our viewModel it can trigger some tracking events that are not related to what we are trying to test. For example if someone will decide to track `userSessionStarted` events. Than `client.events` will contain events that are not part of our testing logic. If we would try to accommodate this "extra" events in our test cases it will make our test not **autonomous**.
By using cont of appearances of specific events, I avoid testing for any side effects that are not part of our testing goal - make test case  **autonomous**.